[FEATURE] Vérifier l'accès à la session de certification des élèves (PIX-1540)

Giroteti opened the pull request:
## :unicorn: Problème
On veut s'assurer qu'un élève qui entre dans une session SCO utilise bien le compte PIX avec lequel il a été réconcilié.

## :robot: Solution
Faire le lien entre le `user` la `schooling-registration` et le `certification-candidate` au moment de l'entrée dans une session de certification SCO.
- Si la `schooling-registration` liée au `user` est aussi liée au `certification-candidate` trouvé par nom/prénom/date de naissance, permettre l'accès
- Sinon, afficher un message d'erreur indiquant à l'utilisateur qu'il est connecté au mauvais compte PIX.  

## :rainbow: Remarques
RAS

## :100: Pour tester
### Nouveau comportement 
- Inscrire deux élèves (certifiables) à une session de certification SCO via l'outil d'inscription dédié au SCO dans Pix-Certif (se connecter avec `certifsco@example.net` et activer le toggle `FT_CERTIF_PRESCRIPTION_SCO`).
Se connecter dans Mon-Pix avec l'élève 1, entrer les données (nom/prénom/ddn) de l'élève 2. On doit voir apparaitre un message indiquant ne pas être connecter avec le bon compte. Entrer ensuite les données correctes de l'élève 1 (nom/prénom/ddn), on doit pouvoir accéder à la session.

### Tests de non-regression 
- Entrer des données invalides (num de session/nom/prenom/ddn). On doit voir apparaître un message générique indiquant qu'on a pas été trouvé.
- Pouvoir accéder à une session non-SCO en entrant session/nom/prenom/ddn

Giroteti committed:
Refacto: move trimming from usecase to controller

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2113.review.pix.fr
- Orga: https://orga-pr2113.review.pix.fr
- Certif: https://certif-pr2113.review.pix.fr
- Admin: https://admin-pr2113.review.pix.fr
- API: https://api-pr2113.review.pix.fr/api/

 Please check it out!

Giroteti committed:
Refacto: move validation error handling to model

Giroteti committed:
Refacto: Candidate link creation via event type

sroccaserra committed:
[sr] Require linkUserToSessionCertificationCandidate usecase in test

sroccaserra committed:
Add a repo function to know if a session is SCO

sroccaserra committed:
Check access for SCO session (student = candidate)

aceol committed:
[SR] link user to session command query

aceol committed:
[SR] Use repository builders in usecase tests

Giroteti committed:
Refacto : turn findXXX into isXXX

in schooling-registration-repository.js :
findByUserIdAndSchoolingRegistrationIdAndSCOOrganization
becomes
isSchoolingRegistrationIdLinkedToUserAndSCOOrganization
because the output is not useful

aceol committed:
Display specific message on student mismatch

Anne-Gaelle-S commented on api/tests/unit/application/session/session-controller_test.js:
```diff
@@ -305,10 +327,13 @@ describe('Unit | Controller | sessionController', () => {
       };
       sinon.stub(certificationCandidateSerializer, 'deserialize').resolves(certificationCandidate);
-      sinon.stub(usecases, 'addCertificationCandidateToSession').withArgs({ sessionId, certificationCandidate }).resolves(addedCertificationCandidate);
+      sinon.stub(usecases, 'addCertificationCandidateToSession').withArgs({
+        sessionId,
+        certificationCandidate,
+      }).resolves(addedCertificationCandidate);
```
**suggestion**
Il me semble qu'ici le withargs ne fonctionnera pas (voir https://nikas.praninskas.com/javascript/2015/07/28/quickie-sinon-withargs-not-working/).
```suggestion
      sinon.stub(usecases, 'addCertificationCandidateToSession');
      usecases.addCertificationCandidateToSession.withArgs({
        sessionId,
        certificationCandidate,
      }).resolves(addedCertificationCandidate);
```

Giroteti replied:
Ben j'ai essayé de modifier le withArgs et le test pète alors que comme ça, il passe donc j'aurais tendance à dire que ça fonctionne.

Anne-Gaelle-S commented on api/tests/unit/application/session/session-controller_test.js:
```diff
@@ -488,55 +513,60 @@ describe('Unit | Controller | sessionController', () => {
-      sinon.stub(certificationCandidateSerializer, 'serialize').withArgs(linkedCertificationCandidate).returns(serializedCertificationCandidate);
+      firstName = 'firstName';
+      lastName = 'lastName';
+      sinon.stub(certificationCandidateSerializer, 'serialize')
+        .withArgs(linkedCertificationCandidate)
+        .returns(serializedCertificationCandidate);
```
**suggestion**
De même ici, pas sûre que le withArgs fonctionne correctement... 
```suggestion
      sinon.stub(certificationCandidateSerializer, 'serialize');
      certificationCandidateSerializer.serialize.withArgs(linkedCertificationCandidate)
        .returns(serializedCertificationCandidate);
```

Giroteti replied:
Idem

Anne-Gaelle-S commented on api/tests/unit/application/session/session-controller_test.js:
```diff
@@ -488,55 +513,60 @@ describe('Unit | Controller | sessionController', () => {
+        .withArgs({ userId, sessionId, firstName: 'firstName', lastName: 'lastName', birthdate }).resolves(new UserAlreadyLinkedEvent());
+      sinon.stub(usecases, 'getCertificationCandidate')
+        .withArgs({ userId, sessionId })
+        .resolves(linkedCertificationCandidate);
```
**suggestion**
Idem pour ces 2 withargs

Anne-Gaelle-S commented on api/tests/unit/application/session/session-controller_test.js:
```diff
@@ -488,55 +513,60 @@ describe('Unit | Controller | sessionController', () => {
+          .resolves(new UserAlreadyLinkedEvent());
+        sinon.stub(usecases, 'getCertificationCandidate')
+          .withArgs({ userId, sessionId })
+          .resolves(linkedCertificationCandidate);
```
**suggestion** 
ici aussi

Anne-Gaelle-S commented on api/tests/unit/application/session/session-controller_test.js:
```diff
@@ -545,20 +575,23 @@ describe('Unit | Controller | sessionController', () => {
+          .resolves(new UserLinkedEvent());
+        sinon.stub(usecases, 'getCertificationCandidate')
+          .withArgs({ userId, sessionId })
+          .resolves(linkedCertificationCandidate);
```
**suggestion**
en fait un peu partout dans le fichier ... 😅

Anne-Gaelle-S commented on api/lib/domain/usecases/link-user-to-session-certification-candidate.js:
```diff
@@ -90,12 +85,40 @@ async function _linkUserToCandidate({
   return certificationCandidate;
 }
+
+class UserLinkedEvent {
+}
+
+class UserAlreadyLinkedEvent {
+}
```
**todo (vu ensemble)**
à supprimer
```suggestion
```

Giroteti replied:
Pas supprimé mais renommé et bougé dans domain/events

Anne-Gaelle-S commented on api/tests/integration/infrastructure/repositories/session-repository_test.js:
```diff
@@ -452,4 +452,36 @@ describe('Integration | Repository | Session', function() {
 
+  describe('#isSco', () => {
+    it('should be true when the certification center is of type SCO', async () => {
+      const certificationCenter = databaseBuilder.factory.buildCertificationCenter({
+        type: 'SCO',
+      });
+      const session = databaseBuilder.factory.buildSession({
+        certificationCenterId: certificationCenter.id,
+      });
+
+      await databaseBuilder.commit();
+
+      const result = await sessionRepository.isSco(session.id);
+
+      expect(result).to.be.true;
+    });
+
+    it('should be false when the certification center is not of type SCO', async () => {
+      const certificationCenter = databaseBuilder.factory.buildCertificationCenter({
+        type: 'PRO',
+      });
+      const session = databaseBuilder.factory.buildSession({
+        certificationCenterId: certificationCenter.id,
+      });
+
+      await databaseBuilder.commit();
+
+      const result = await sessionRepository.isSco(session.id);
+
+      expect(result).to.be.false;
```
**suggestion (non bloquante)**
Il manque juste les given, when, then
```suggestion
      // given
      const certificationCenter = databaseBuilder.factory.buildCertificationCenter({
        type: 'SCO',
      });
      const session = databaseBuilder.factory.buildSession({
        certificationCenterId: certificationCenter.id,
      });

      await databaseBuilder.commit();

      // when
      const result = await sessionRepository.isSco(session.id);

      // then
      expect(result).to.be.true;
    });

    it('should be false when the certification center is not of type SCO', async () => {
      // given
      const certificationCenter = databaseBuilder.factory.buildCertificationCenter({
        type: 'PRO',
      });
      const session = databaseBuilder.factory.buildSession({
        certificationCenterId: certificationCenter.id,
      });

      await databaseBuilder.commit();

      // when
      const result = await sessionRepository.isSco(session.id);

      // then
      expect(result).to.be.false;
```

Anne-Gaelle-S commented on api/tests/unit/domain/usecases/link-user-to-session-certification-candidate_test.js:
```diff
@@ -230,61 +257,275 @@ describe('Unit | Domain | Use Cases | link-user-to-session-certification-candid
-          sinon.stub(certificationCandidateRepository, 'findOneBySessionIdAndUserId')
-            .withArgs({ sessionId, userId })
-            .resolves(undefined);
+        expect(err).to.be.instanceOf(CertificationCandidateByPersonalInfoNotFoundError);
```
**suggestion (non bloquant)**
Ce suspence ... given, when .... eeeettt .... suussspeennnce
```suggestion
        // then
        expect(err).to.be.instanceOf(CertificationCandidateByPersonalInfoNotFoundError);
```

Anne-Gaelle-S commented on api/tests/unit/domain/usecases/link-user-to-session-certification-candidate_test.js:
```diff
@@ -230,61 +257,275 @@ describe('Unit | Domain | Use Cases | link-user-to-session-certification-candid
+          schoolingRegistrationRepository,
+          sessionRepository,
+        });
+
+        expect(err).to.be.instanceOf(MatchingReconciledStudentNotFoundError);
```
**suggestion (non bloquant)**
```suggestion
        
        // then
        expect(err).to.be.instanceOf(MatchingReconciledStudentNotFoundError);
```

Anne-Gaelle-S commented on mon-pix/app/components/certification-joiner.js:
```diff
@@ -6,6 +6,7 @@ import Component from '@ember/component';
 import { action } from '@ember/object';
 import { inject as service } from '@ember/service';
 import { tracked } from '@glimmer/tracking';
+import _get from 'lodash/get';
```
**question**
C'est vraiment `_get` ou c'est `get` qu'il faut ici ?
Si c'est bien `_get` ça correspond à quoi / y'a une différence avec `get` ?

Giroteti replied:
j'ai l'impression que ça sert à indiquer que ça provient de lodash mais sans importer tout lodash (auquel cas on ferait ça `_.get`). C'est ça @aceol ?

Giroteti committed:
Refacto : move events to /domain/events

octo-topi commented:
je ne vois pas de définition de la variable d'env  FT_CERTIF_PRESCRIPTION_SCO dans la doc :cry: 
- api/sample.env
- https://1024pix.atlassian.net/wiki/spaces/DEV/pages/957677602/2019-07-16+Glossaire+des+variables+d+environnement

Giroteti commented:
> je ne vois pas de définition de la variable d'env FT_CERTIF_PRESCRIPTION_SCO dans la doc 😢
> 
> * api/sample.env
> * https://1024pix.atlassian.net/wiki/spaces/DEV/pages/957677602/2019-07-16+Glossaire+des+variables+d+environnement

Décision : Le toggle étant voué à disparaître très prochainement (un ticket JIRA est d'ores et déjà prévu à cet effet) et documenté via les tests, on choisit de ne pas le documenter ailleurs.

Giroteti committed:
Remove unused repository stub in tests
